### PR TITLE
chore: vhs utils update

### PR DESCRIPTION
**Pull request by brandonocasey**
Requires the following:
~~https://github.com/videojs/aes-decrypter/pull/81~~
~~https://github.com/videojs/m3u8-parser/pull/118~~
~~https://github.com/videojs/mpd-parser/pull/115~~

### CODE REVIEW

**Review by gkatsev — DISMISSED** [review]
Works well, just needs to wait for the vhs-utils update.

File: src/util/codecs.js [hunk 1]
```diff
@@ -55,6 +55,38 @@ export const isMuxed = (master, media) => {
   return false;
 };
 
+export const unwrapCodecList = function(codecList) {
+  const codecs = {};
+
+  codecList.forEach(({mediaType, type, details}) => {
+
+    // TODO: log a warning, something like:
```
Comment (gesinger) [inline src/util/codecs.js:63]: Should we add this log?

**Review by gkatsev — DISMISSED** [review]
Did a test and it's working. Confirmed that with this change we will properly pick up missing codecs and be able to play them even if they are partially missing from the manifest.

**Review by gkatsev — DISMISSED** [review]

**Review by gesinger — DISMISSED** [review]

File: test/videojs-http-streaming.test.js [hunk 2]
```diff
@@ -1805,6 +1805,9 @@ QUnit.test('blacklists incompatible playlists by codec, without codec switching'
 });
 
 QUnit.test('does not blacklist incompatible codecs with codec switching', function(assert) {
+  const oldIsTypeSupported = window.MediaSource.isTypeSupported;
```
Comment (brandonocasey) [inline test/videojs-http-streaming.test.js:1808]: edge support ac-3, we need to mock that it doesn't support it for this test.

File: src/master-playlist-controller.js [hunk 3]
```diff
@@ -1677,67 +1675,61 @@ export class MasterPlaylistController extends videojs.EventTarget {
...
-          variantCodecs.audio.type.toLowerCase() !== codecs.audio.type.toLowerCase()) {
-          variant.excludeUntil = Infinity;
-          blacklistReasons.push(`audio codec "${variantCodecs.audio.type}" !== "${codecs.audio.type}"`);
+        if (variantAudioDetails && audioDetails && variantAudioDetails.type.toLowerCase() !== audioDetails.type.toLowerCase()) {
+          blacklistReasons.push(`audio codec "${variantAudioDetails.type}" !== "${audioDetails.type}"`);
         }
       }
 
       if (blacklistReasons.length) {
+        variant.excludeUntil = Infinity;
```
Comment (brandonocasey) [inline src/master-playlist-controller.js:1732]: moved the exclude down here so we don't do it in every if.

File: src/master-playlist-controller.js [hunk 4]
```diff
@@ -1677,67 +1675,61 @@ export class MasterPlaylistController extends videojs.EventTarget {
...
-    const codecs = parseCodecs(codecString);
-    const codecCount = Object.keys(codecs).length;
+    const ids = [];
+    const playlists = this.master().playlists;
+    const codecs = unwrapCodecList(parseCodecs(codecString));
+    const codecCount_ = codecCount(codecs);
+    const videoDetails = codecs.video && parseCodecs(codecs.video)[0] || null;
+    const audioDetails = codecs.audio && parseCodecs(codecs.audio)[0] || null;
+
+    Object.keys(playlists).forEach((key) => {
```
Comment (brandonocasey) [inline src/master-playlist-controller.js:1685]: loop through all playlists for sure

File: test/videojs-http-streaming.test.js [hunk 5]
```diff
@@ -1940,6 +1940,9 @@ QUnit.test('blacklists fmp4 playlists by browser support', function(assert) {
 });
 
 QUnit.test('blacklists ts playlists by muxer support', function(assert) {
+  const oldIsTypeSupported = window.MediaSource.isTypeSupported;
```
Comment (brandonocasey) [inline test/videojs-http-streaming.test.js:1947]: again with browser codec support needing to be overridden.

**Review by gesinger — APPROVED** [review]

**Pull request merged**